Copy a file to the same directory

greenido opened the pull request:
1. Added the logic to change the new file name to 'xxxx_1' (or 2,3,4, when there are already copies of the file).
2. After the copy there was a need to reload the list of files in order to show the user the changes.


Fixes #9931

Commit:
fixing issue 9931 - copy a file to the same directory

Signed-off-by: Ido Green <greenido@gmail.com>
Signed-off-by: Morris Jobke <hey@morrisjobke.de>

skjnldsv commented on .gitignore:
```diff
@@ -149,3 +149,4 @@ clover.xml
 # Tests - dependencies
 tests/acceptance/composer.lock
 tests/acceptance/vendor/
+
```
Not needed

skjnldsv commented on .htaccess:
```diff
@@ -78,3 +78,7 @@ Options -Indexes
 <IfModule pagespeed_module>
   ModPagespeed Off
 </IfModule>
+#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####
```
Please remove those changes

greenido replied:
ok

MorrisJobke commented:
@skjnldsv Let's make it 15 and then backport to  stable14 😉

Commit:
cleaning after comments on issue 10825

More: https://github.com/nextcloud/server/pull/10825#issuecomment-415800850

Signed-off-by: Ido Green <greenido@gmail.com>

greenido commented:
The changes @skjnldsv suggested are in the last commit: https://github.com/nextcloud/server/pull/10825/commits/83bd2f5571b37cc1d5d1303c18a5bebdc6519274
Thx.

MorrisJobke commented:
@dairaine @sunjam @lightweight Could you try this change for the feature you requested?

jancborchardt commented:
@greenido nice! Could you adjust the naming to the spec at https://github.com/nextcloud/server/issues/9931#issuecomment-402318120 ?

> - filename (copy).txt
> - filename (another copy).txt
> - filename (3rd copy).txt
> - filename (4th copy).txt
> - etc
> 
> This is more obvious than just appending the number, which could also be a conflict or having uploaded the original file again.

jancborchardt reviewed (changes requested):
@greenido nice! Could you adjust the naming to the spec at https://github.com/nextcloud/server/issues/9931#issuecomment-402318120 ?

> - filename (copy).txt
> - filename (another copy).txt
> - filename (3rd copy).txt
> - filename (4th copy).txt
> - etc
> 
> This is more obvious than just appending the number, which could also be a conflict or having uploaded the original file again.

greenido commented:
Sure! I hope to find time tonight ;)


On Tue, Oct 2, 2018 at 10:17 AM Jan-Christoph Borchardt <
notifications@github.com> wrote:

> *@jancborchardt* requested changes on this pull request.
>
> @greenido <https://github.com/greenido> nice! Could you adjust the naming
> to the spec at #9931 (comment)
> <https://github.com/nextcloud/server/issues/9931#issuecomment-402318120> ?
>
>
>    - filename (copy).txt
>    - filename (another copy).txt
>    - filename (3rd copy).txt
>    - filename (4th copy).txt
>    - etc
>
> This is more obvious than just appending the number, which could also be a
> conflict or having uploaded the original file again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nextcloud/server/pull/10825#pullrequestreview-160831406>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADwIC1SM02XTolvwzw6gRTVmAWcJ6dKks5ug5_DgaJpZM4WJ9Z1>
> .
>

Commit:
Use the new naming per the comments on #9931

More details here: https://github.com/nextcloud/server/issues/9931#issuecomment-402318120

Signed-off-by: Ido Green <greenido@gmail.com>

greenido commented:
Done.
Please check and let me know. Thank you.

On Tue, Oct 2, 2018 at 10:42 AM Ido Green <greenido@gmail.com> wrote:

> Sure! I hope to find time tonight ;)
>
>
> On Tue, Oct 2, 2018 at 10:17 AM Jan-Christoph Borchardt <
> notifications@github.com> wrote:
>
>> *@jancborchardt* requested changes on this pull request.
>>
>> @greenido <https://github.com/greenido> nice! Could you adjust the
>> naming to the spec at #9931 (comment)
>> <https://github.com/nextcloud/server/issues/9931#issuecomment-402318120>
>> ?
>>
>>
>>    - filename (copy).txt
>>    - filename (another copy).txt
>>    - filename (3rd copy).txt
>>    - filename (4th copy).txt
>>    - etc
>>
>> This is more obvious than just appending the number, which could also be
>> a conflict or having uploaded the original file again.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/nextcloud/server/pull/10825#pullrequestreview-160831406>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AADwIC1SM02XTolvwzw6gRTVmAWcJ6dKks5ug5_DgaJpZM4WJ9Z1>
>> .
>>
>

skjnldsv commented on apps/files/js/filelist.js:
```diff
@@ -2191,7 +2191,40 @@
+							let maxNum = 1;
+							if (self.files !== null) {
+								leftPartOfName = leftPartOfName.replace("/", "");
+								leftPartOfName = leftPartOfName.replace(/\(copy \d+\)/,"");
```
Will \d works with foreign char like Japan or China?

greenido replied:
I think so... but don't have these lang. on my machine.
Can you please check?

skjnldsv commented on apps/files/js/filelist.js:
```diff
@@ -2191,7 +2191,40 @@
+						let fileNumber = leftPartOfName.match(/\d+/);
+						if (isNaN(fileNumber) ) {
+							fileNumber++;
+							targetPathAndName = targetPathAndName.replace(/(?=\.[^.]+$)/g, " (copy " + fileNumber + ")");	
```
We should use a translated string for that so that the renaming also display the same string than the selected language.

greenido replied:
Good catch! 
Can you show me where to look for usage of translated strings?

skjnldsv replied:
Ah sorry, I missed your comment!
You can use the `t` function. 

e.g. 
``` js
						t('files', '{loadedSize} of {totalSize} ({bitrate})' , {
							loadedSize: humanFileSize(data.loaded),
							totalSize: humanFileSize(data.total),
							bitrate: humanFileSize(data.bitrate / 8) + '/s'
						})
```

skjnldsv reviewed (dismissed):
Need some small changes!! Did not tested it yet, the code looks good 👍✨

jancborchardt reviewed (changes requested):
The first copy does not need a number, just `Filename (copy).txt` :) Otherwise it seems good. :+1:

Commit:
Per the last comment on having only (copy) for the first one

Signed-off-by: Ido Green <greenido@gmail.com>

MorrisJobke commented:
@jancborchardt @skjnldsv Mind to re-review?

greenido commented:
As #2 - without a refresh I the view is not showing the new copied file.
Am I missing something?

On Wed, Nov 7, 2018 at 10:26 AM Jan-Christoph Borchardt <
notifications@github.com> wrote:

> It seems good. But 2 things:
>
>    - Why the notification "Copied filename.jpg inside"? That’s unnecessary
>    - Why does the view refresh?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nextcloud/server/pull/10825#issuecomment-436727733>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADwIFpc_V52fp7eZrb8QpQaTT_k5sUbks5usyXjgaJpZM4WJ9Z1>
> .
>

jancborchardt commented:
@greenido right, in that case a :+1: from me! :)

skjnldsv commented:
Still untranslated :)

MorrisJobke commented:
> Still untranslated :)

That's fine as long as it goes in today we can ship this one. I plan to review it now :)

MorrisJobke commented:
> That's fine as long as it goes in today we can ship this one. I plan to review it now :)

Ah - now I got it.

MorrisJobke committed:
Translate name for "(copy)"

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke reviewed (approved):
I fixed a variable name and changed the occurence of `copy` to the translated version. Tested and works here 👍

MorrisJobke commented:
@greenido As we are planning to release 15.0.0 Beta 2 today: I guess there is very little chance that you could sign-off the commits, right? It's fairly easy: https://github.com/nextcloud/server/pull/10825/checks?check_run_id=32106619

MorrisJobke commented:
> @greenido As we are planning to release 15.0.0 Beta 2 today: I guess there is very little chance that you could sign-off the commits, right? It's fairly easy:

Let's do this for 16 and look into back porting this maybe later.

greenido commented:
I'll do my best to do it today - Thank you for the heads up.

On Thu, Nov 15, 2018 at 7:42 AM Morris Jobke <notifications@github.com>
wrote:

> @greenido <https://github.com/greenido> As we are planning to release
> 15.0.0 Beta 2 today: I guess there is very little chance that you could
> sign-off the commits, right? It's fairly easy:
>
> Let's do this for 16 and look into back porting this maybe later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nextcloud/server/pull/10825#issuecomment-439084673>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADwICDxQMf3MdqphIfbi8sgs9-6PmJ2ks5uvYtvgaJpZM4WJ9Z1>
> .
>

MorrisJobke commented:
@greenido No need to hurry on this one. Beta 2 is out 🚀 but we will get it into the next release maybe ;)

greenido reviewed (commented):
Signed-off-by: Ido Green <greenido@gmail.com>

danielkesselberg commented:
@greenido https://github.com/nextcloud/server/runs/32106619 follow the instructions there to fix the sign off

greenido commented:
I got into a small mess due to an update I did to my fork...
So the two git commands aren’t working as expected.


> On Nov 15, 2018, at 09:36, Daniel Kesselberg <notifications@github.com> wrote:
> 
> @greenido https://github.com/nextcloud/server/runs/32106619 follow the instructions there to fix the sign off
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

danielkesselberg commented:
Yes. I'm not sure but greenido/server is missing a lof ot changes form nextcloud/server. I would try to update my fork first (https://help.github.com/articles/syncing-a-fork/) and than try to rebase again.

greenido commented:
Just did... thank you for the heads up.

--

> On Nov 15, 2018, at 10:49, Daniel Kesselberg <notifications@github.com> wrote:
> 
> Yes. I'm not sure but greenido/server is missing a lof ot changes form nextcloud/server. I would try to update my fork first (https://help.github.com/articles/syncing-a-fork/) and than try to rebase again.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

MorrisJobke commented:
> Just did... thank you for the heads up.

This looks like it went wrong 🙈 Sorry for the mess here. I pushed to your branch a new commit as well and this seems to not have relaxed the situation.

I guess here a heavy "git rebase --interactive" is needed with "edit" as option instead of "pick" to be able to modify the commit message (adding basically the sign-off message).  See https://morrisjobke.de/2015/12/03/How-to-do-a-git-rebase/ for details about a rebase.

In your case it's best to checkout master. Do a pull from our master. Then you can do the above mentioned rebase and push again. In case of problems: I have the old version of the branch still locally and we can go for that if something bad happens.

greenido commented:
Hey @MorrisJobke - I followed the steps you mention but I don't know what to pick in the interactive merge... 
See:
==
Using index info to reconstruct a base tree...
M	core/js/setupchecks.js
M	settings/Controller/CheckSetupController.php
Falling back to patching base and 3-way merge...
Auto-merging settings/Controller/CheckSetupController.php
CONFLICT (content): Merge conflict in settings/Controller/CheckSetupController.php
Auto-merging core/js/setupchecks.js
CONFLICT (content): Merge conflict in core/js/setupchecks.js
error: Failed to merge in the changes.
Patch failed at 0043 Adds a setup check for the memory limit
==
Can you fix these two in 'my' local master (or branch)?
Thanks!

MorrisJobke commented:
I just used the local commits from my side, rebased them and pushed them. They do not yet contain the sign-off message, but I can walk you through this.

```
git rebase --interactive def8af3e7e
```

This will open an editor with following content:

```
pick 0dad9c1397 fixing issue 9931 - copy a file to the same directory
pick 1d827da7e2 cleaning after comments on issue 10825
pick 8eab3e8cb1 Use the new naming per the comments on #9931
pick d025c92108 Per the last comment on having only (copy) for the first one
pick bdf3019e4e Translate name for "(copy)"

# Rebase def8af3e7e..bdf3019e4e onto def8af3e7e (5 commands)
#
# Commands:
# p, pick = use commit
# r, reword = use commit, but edit the commit message
# e, edit = use commit, but stop for amending
# s, squash = use commit, but meld into previous commit
# f, fixup = like "squash", but discard this commit's log message
# x, exec = run command (the rest of the line) using shell
# d, drop = remove commit
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out
``` 

The first 4 lines in there are your commits, the fifth one is my commit. So you only need to edit the first 4 lines. Change there the `pick` into `edit`. This basically does a pick and gives you a prompt after each of those commits. Then you can run `git commit --amend -s` to add the sign-off message. Continue then to the next commit with `git rebase --continue`. Do there the same. Repeat this until it says:

```
Successfully rebased and updated refs/heads/fixing-issue-9931.
```

Force push your commits and you should be done. 👍

MorrisJobke commented:
Uuuuhhh ... the git history looks a bit broken. It seems that there were non-fast-forward merges on your master branch and thus git thinks that there needs to be more merged into this one here. Also your commits do not contain the message. :/ I could add the message for you and fix this PR. In the future you could use `git commit -s` to let git add the message right while creating the commit and this then does not require this huge rebasing here anymore.

MorrisJobke committed:
JSUnit tests are not compatible with let/const yet

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
stable15 is branched off -> merge

greenido commented:
Will do! 
Thank you 🙏 

--

> On Nov 22, 2018, at 11:37, welcome[bot] <notifications@github.com> wrote:
> 
> Thanks for your first pull request and welcome to the community! Feel free to keep them coming! If you are looking for issues to tackle then have a look at this selection: https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22
> Most developers hang out on IRC. So join #nextcloud-dev on Freenode for a chat!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
